docs/design.md

```diff
@@ -0,0 +1,326 @@
+// ID is the internal ID that the worker uses to identify an Exec
+// Under the hood this might just be a UUID
+type ID string
+
+type ResourceLimits struct {
+	MaxMemoryBytes int64
+	// ...
+}
+
+// ProcessRequest represents a request to execute a Linux process in the worker
+type ProcessRequest struct {
+	// Command is the command executed to be executed together with its arguments
+	Command string
+
+	// ResourceLimits specifies the resources that the process will have access to
+	// These translate to cgroup interface files configuration.
+	ResourceLimits ResourceLimits
+
+	RequestedBy string
+}
+
+// StartProcess starts a new process and adds it to the worker
+// process registry. It does not wait for the process to terminate.
+func (w *Worker) StartProcess(req ProcessRequest) (ID, error) {
+	// ...
+}
+
+// StopProcess stops a process currently managed by the worker
+// Returns an error if there was any issues stopping the process
+// or the process does not exist in the worker registry.
+func (w *Worker) StopProcess(processId ID) error {
+	// ...
+}
+
+// ProcessInfo gives access to information about a
+// process.
+type ProcessInfo interface {
+	PID() int
+	StartedBy() string
+	State() string
+	StartedAt() time.Time
+	FinishedAt() time.Time
+}
+
+// GetProcessInfo gives access to the ProcessInfo interface
+// which allows querying for information about the process.
+func (w *Worker) GetProcessInfo(processId ID) ProcessInfo {
+	// ...
+}
+
+// StreamProcessOutput returns an instance of a Streamer that
+// can be used to stream the combined stdout and stderr of
+// a process managed by the worker
+func (w *Worker) StreamProcessOutput(processId ID) Streamer {
+	// ...
+}
+
+type ProcessOutputEntry struct {
+	Content    []byte
+	ReceivedAt time.Time
+}
+
+// Streamer is an interface used to stream output from a Process
+// managed by the worker
+type Streamer interface {
+	// StreamProcessOuput returns a channel to receive process output
+	// in real time. If any error are encountered during execution the error
+	// will be sent in the errorChan. Otherwise, outputChan only closes when
+	// process is done sending output
+	StreamProcessOutput() (outputChan <-chan ProcessOutputEntry, errChan <-chan error)
+}
+```
+
+### Output Streaming
+
+The following interface is used for output streaming:
+
+```go
+// Streamer is an interface used to stream output from a Process
+// managed by the worker
+type Streamer interface {
+	// StreamProcessOuput returns a channel to receive process output
+	// in real time. If any error are encountered during execution the error
+	// will be sent in the errorChan. Otherwise, outputChan only closes when
+	// process is done sending output
+	StreamProcessOutput() (outputChan <-chan ProcessOutputEntry, errChan <-chan error)
+}
+```
+
+* When a process is started, a new goroutine is spanwed that buffers output
+and forwards to listeners. 
+* When a new Streamer is created, `outputChan` first receives output up to the
```

Comment:
Is there scope for output to be dropped in this transition from playing catchup to live streaming output? How will you manage that transition so this doesn't happen?

Reply:
Added some additional information about how we will prevent this.